src/coreclr/gc/gcpriv.h

```diff
@@ -4299,142 +4489,472 @@ class gc_heap
+
+                dprintf (6666, ("hc would change %.3f, factor is %d", hc_change_pct, hc_change_factor));
+                if (hc_change_pct < 0.2)
+                {
+                    // Should we also consider absolute time here?
+                    int delayed_hc_change_factor = hc_change_factor * 3;
+                    int count = 0;
+                    if (adj->metric == adjust_budget)
+                    {
+                        count = adj->count;
+                    }
+
+                    dprintf (6666, ("we've changed budget instead of HC %d times from %Id GCs ago, thres %d times",
+                                    count, num_gcs_since_change, delayed_hc_change_factor));
+
+                    if (count < delayed_hc_change_factor)
+                    {
+                        adj_metric = adjust_budget;
+                    }
+                }
+                else
+                {
+                    bool change_p = (num_gcs_since_change > (size_t)(hc_change_factor * sample_size));
+                    dprintf (6666, ("It's been %Id GCs since we wanted to change HC last time, thres %d GCs, %s",
+                        num_gcs_since_change, (hc_change_factor * sample_size), (change_p ? "change" : "don't change yet")));
+                    if (!change_p)
+                    {
+                        adj_metric = not_adjusted;
+                    }
+                }
+            }
+
+            dprintf (6666, ("conclusion: %s", str_adjust_metrics[adj_metric]));
+            if (adj_metric == adjust_hc)
+            {
+                reset_budget_adjustment();
+            }
+            return adj_metric;
         }
 
+        //
+        // Budget computation. Now we have 2 budgets -
+        // BCS (Budget Computed via Survrate) and
+        // BCD (Budget Computed via DATAS)
+        //
+        // When DATAS is on, BCD is our upper bound because we want to control how much allocation to allow before
+        // the next GC happens which directly contributes to the heap size.
+        // If BCS is smaller, it means we have room to adjust budget. Currently we don't adjust it dramatically because
+        // then we risk affecting tcp too much (tcp is still calculated using both the collection time and the wait
+        // time in msl).
+        //
+
         size_t          max_gen0_new_allocation;
         size_t          min_gen0_new_allocation;
 
-        size_t compute_gen0_new_allocation (size_t total_old_gen_size)
+        size_t compute_total_gen0_budget (size_t total_soh_stable_size)
         {
-            assert (total_old_gen_size > 0);
+            assert (total_soh_stable_size > 0);
 
-            // TODO: adjust these based on conserve_mem_setting.
-            double old_gen_growth_factor = 16.0 / sqrt ((double)total_old_gen_size / 1000.0 / 1000.0);
+            float factor = (float)(20 - conserve_mem_setting);
```

Comment:
Is it worth adding a comment explaining why 20 is used here?

Reply:
20 was chosen to make it close to baseline and enabling this to be adjusted by the conserve memory setting (20 - 5 = 15, close to 16). this 20 may change. I haven't put in much effort to document these numbers because they might change. I still consider this feature heavily under development.